// @smartthings/core-sdk/dist/endpoint-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const qs_1 = __importDefault(require("qs"));
const logger_1 = require("./logger");
exports.defaultSmartThingsURLProvider = {
    baseURL: 'https://api.smartthings.com',
    authURL: 'https://auth-global.api.smartthings.com/oauth/token',
    keyApiURL: 'https://key.smartthings.com',
};
class EndpointClient {
    constructor(basePath, config) {
        this.basePath = basePath;
        this.config = config;
        this.logger = config.logger ? config.logger : logger_1.noLogLogger;
    }
    setHeader(name, value) {
        if (!this.config.headers) {
            this.config.headers = {};
        }
        this.config.headers[name] = value;
        return this;
    }
    removeHeader(name) {
        if (this.config.headers) {
            delete this.config.headers[name];
        }
        return this;
    }
    url(path) {
        var _a, _b, _c;
        if (path) {
            if (path.startsWith('/')) {
                return `${(_a = this.config.urlProvider) === null || _a === void 0 ? void 0 : _a.baseURL}${path}`;
            }
            else if (path.startsWith('https://')) {
                return path;
            }
            return `${(_b = this.config.urlProvider) === null || _b === void 0 ? void 0 : _b.baseURL}/${this.basePath}/${path}`;
        }
        return `${(_c = this.config.urlProvider) === null || _c === void 0 ? void 0 : _c.baseURL}/${this.basePath}`;
    }
    /* eslint-disable @typescript-eslint/no-explicit-any */
    request(method, path, data, params) {
        return __awaiter(this, void 0, void 0, function* () {
            const headers = this.config.headers ? Object.assign({}, this.config.headers) : {};
            if (this.config.loggingId) {
                headers['X-ST-CORRELATION'] = this.config.loggingId;
            }
            if (this.config.version) {
                const versionString = `application/vnd.smartthings+json;v=${this.config.version}`;
                // Prepare the accept header
                if (typeof headers.Accept === 'undefined' || headers.Accept === 'application/json') {
                    headers.Accept = versionString;
                }
                else {
                    headers.Accept = `${versionString}, ${headers.Accept}`;
                }
            }
            let axiosConfig = {
                url: this.url(path),
                method,
                headers,
                params,
                data,
                paramsSerializer: params => qs_1.default.stringify(params, { indices: false }),
            };
            axiosConfig = yield this.config.authenticator.authenticate(axiosConfig);
            if (this.logger.isDebugEnabled()) {
                this.logger.debug(`making axios request: ${JSON.stringify(axiosConfig)}`);
            }
            let response;
            try {
                response = yield axios_1.default.request(axiosConfig);
                return response.data;
            }
            catch (error) {
                if (error.response && error.response.status === 401 && this.config.authenticator.refresh) {
                    if (this.config.authenticator.acquireRefreshMutex) {
                        const release = yield this.config.authenticator.acquireRefreshMutex();
                        try {
                            yield this.config.authenticator.refresh(axiosConfig, this.config);
                            response = yield axios_1.default.request(axiosConfig);
                            return response.data;
                        }
                        finally {
                            release();
                        }
                    }
                    else {
                        yield this.config.authenticator.refresh(axiosConfig, this.config);
                        response = yield axios_1.default.request(axiosConfig);
                        return response.data;
                    }
                }
                // Annotate message with SmartThings API error data
                if (error.response && error.response.data) {
                    error.message = error.message + ': ' + JSON.stringify(error.response.data);
                }
                throw error;
            }
        });
    }
    get(path, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.request('get', path, undefined, params);
        });
    }
    post(path, data, params) {
        return this.request('post', path, data, params);
    }
    put(path, data, params) {
        return this.request('put', path, data, params);
    }
    patch(path, data, params) {
        return this.request('patch', path, data, params);
    }
    delete(path, params) {
        return this.request('delete', path, undefined, params);
    }
    getPagedItems(path, params) {
        return __awaiter(this, void 0, void 0, function* () {
            let list = yield this.get(path, params);
            const result = list.items;
            while (list._links && list._links.next) {
                list = yield this.get(list._links.next.href);
                result.push(...list.items);
            }
            return result;
        });
    }
}
exports.default = EndpointClient;
