// @smartthings/core-sdk/dist/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * A simple implementation of the Logger interface that does not log anything.
 */
class NoLogLogger {
    constructor() {
        this.level = 'off';
    }
    /* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars */
    log(...args) {
        // no-op
    }
    trace(message, ...args) {
        // no-op
    }
    debug(message, ...args) {
        // no-op
    }
    info(message, ...args) {
        // no-op
    }
    warn(message, ...args) {
        // no-op
    }
    error(message, ...args) {
        // no-op
    }
    fatal(message, ...args) {
        // no-op
    }
    /* eslint-enable */
    isTraceEnabled() {
        return false;
    }
    isDebugEnabled() {
        return false;
    }
    isInfoEnabled() {
        return false;
    }
    isWarnEnabled() {
        return false;
    }
    isErrorEnabled() {
        return false;
    }
    isFatalEnabled() {
        return false;
    }
}
exports.NoLogLogger = NoLogLogger;
exports.noLogLogger = new NoLogLogger();
