// @smartthings/core-sdk/dist/endpoint/deviceprofiles.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = __importDefault(require("../endpoint-client"));
const types_1 = require("../types");
var DeviceProfileStatus;
(function (DeviceProfileStatus) {
    DeviceProfileStatus["DEVELOPMENT"] = "DEVELOPMENT";
    DeviceProfileStatus["PUBLISHED"] = "PUBLISHED";
})(DeviceProfileStatus = exports.DeviceProfileStatus || (exports.DeviceProfileStatus = {}));
class DeviceProfilesEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.default('deviceprofiles', config));
    }
    /**
     * List all the device profiles belonging to the principal (i.e. user)
     */
    list() {
        return this.client.getPagedItems();
    }
    /**
     * Get the definition of a specific device profile
     * @param id UUID of the device profile
     */
    get(id) {
        return this.client.get(id);
    }
    /**
     * Delete a device profile
     * @param id UUID of the device profile
     */
    delete(id) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.delete(id);
            return types_1.SuccessStatusValue;
        });
    }
    /**
     * Create a device profile
     * @param data device profile definition
     */
    create(data) {
        return this.client.post(undefined, data);
    }
    /**
     * Update a device profile
     * @param id UUID of the device profile
     * @param data the new device profile definition
     */
    update(id, data) {
        return this.client.put(id, data);
    }
    /**
     * Update the status of a device profile
     * @param id UUID of the device profile
     * @param deviceProfileStatus new device profile status
     */
    updateStatus(id, deviceProfileStatus) {
        return this.client.post(`${id}/status`, { deviceProfileStatus });
    }
}
exports.DeviceProfilesEndpoint = DeviceProfilesEndpoint;
